README.md
# Unit Conversion Library [![NPM package][npm-image]][npm-url] [![Build Status][ci-image]][ci-url] [![Coverage][coverage-image]][coverage-url]

Convert between distance, pressure, temperature, volume and mass units easily!

# Installation

`$ npm install smartcar-unit`

# Example Usage

```javascript
var unit = require('smartcar-unit');
var miles = unit(184000, 'miles');
var kilometers = miles.as('kilometers'); // 296119.296
var kilometersRounded = miles.as('kilometers', 2); // 296119.3
```

# Supported Units

| Distance | aliases |
|----------|---------|
| inch | `in` `inch` `inches`|
| foot | `ft` `foot` `feet` |
| mile | `mi` `mile` `miles` |
| meter | `m` `meter` `meters` `metre` `metres` |
| kilometer | `km` `kilometer` `kilometers` |

| Pressure | aliases |
|----------|---------|
| pounds per square inch| `psi` `psis` |
| pascal | `pa` `pascal` `pascals` |
| bar | `bar` `bars` |
| millibar | `mbar` `millibar` `mbars` `millibars` `hectopascal` `hectopascals` |
| millimeter of mercury | `mmhg` `torr` |
| atmospheric | `atmosphere` `atmospheres` `atmospheric` `barometric` |
| kilopascal | `kpa` `kpas` `kilopascal` `kilopascals` |

| Temperature | aliases |
| ----------- | ------- |
| kelvin | `k` `kelvin` `kelvins` |
| fahrenheit | `f` `fahrenheit` |
| celsius | `c` `celsius` `centigrade` |

| Mass | aliases |
| ---- | ------- |
| ounce | `oz` `ounce` `ounces` |
| pound | `lb` `lbs` `pound` `pounds` |
| gram | `g` `gm` `gram` `grams` |
| kilogram | `kg` `kilogram` `kilograms` |

| Volume | aliases |
| ------ | ------- |
| milliliter  | `ml` `milliliter` `milliliters` `millilitre` `millilitres`|
| liter  | `l` `liter` `liters` `litre` `litres`|
| us gallon | `usgal` `usgallon` `usgallons` |
| imperial gallon | `gal` `gallon` `gallons` |

[npm-url]: https://www.npmjs.com/package/smartcar-unit
[npm-image]: https://img.shields.io/npm/v/smartcar-unit.svg?style=flat-square

[ci-url]: https://travis-ci.org/smartcar/unit
[ci-image]: https://img.shields.io/travis/smartcar/unit/master.svg?style=flat-square

[coverage-url]: https://codecov.io/gh/smartcar/unit
[coverage-image]: https://img.shields.io/codecov/c/github/smartcar/unit/master.svg?style=flat-square


CHANGELOG.md
# 1.1.1 (2017-03-30)

## Features
* Add volume conversions

## Meta
* [Comparison since last release](https://github.com/smartcar/unit/compare/v1.1.0...v1.1.1)

# 1.1.0 (2016-12-15)

Mostly an internal code reorganization, only external changes are a typo fix
for `celsius` and the addition of `centimeters`

From this point forward node version older than 4.0.0  will no longer be officially supported

## Features
* Add centimeters
* Travis builds will be cached for faster building

## Bug Fixes
* Fixed typo `celcius -> celsius`

## Meta
* [Comparison since last release](https://github.com/smartcar/unit/compare/v1.0.2...v1.1.0)

# 1.0.2 (2016-07-13)

## Features
* Add units of mass

## Meta
* [Comparison since last release](https://github.com/smartcar/unit/compare/v1.0.1...v1.0.2)

# 1.0.1 (2016-07-08)

## Features
* Control the precision of the number reported back
* Add build, coverage and npm badges to readme

## Bug Fixes
* Documentation referred to package as `unit`, renamed to `smartcar-unit`
* Fix the ratios for `in`, `feet` and `miles`

## Meta
* [Comparison since last release](https://github.com/smartcar/unit/compare/v1.0.0...v1.0.1)

# 1.0.0 (2016-07-08)

Initial release


test/index.js
'use strict';

const test = require('ava');
const unit = require('../index');

function approximately(value, expected) {
  return Math.abs(value - expected) <= 0.0001;
}

test('in, as, and to syntaxes are equivalent', function(t) {
  var inches = unit(12, 'inches');
  var toSyntax = inches.to('feet');
  var asSyntax = inches.as('feet');
  var inSyntax = inches.in('feet');
  t.is(toSyntax, inSyntax);
  t.is(toSyntax, asSyntax);
});

test('rounding', function(t) {
  var rounded = unit(13, 'inches').to('feet', 2);
  t.is(rounded, 1.08);
});

test('inches', function(t) {
  var km = unit(100000, 'inches').as('km');
  t.true(approximately(km, 2.54));
});

test('feet', function(t) {
  var km = unit(10000, 'feet').as('km');
  t.true(approximately(km, 3.048));
});

test('miles', function(t) {
  var km = unit(184000, 'miles').as('kilometers');
  t.true(approximately(km, 296119.296));
});

test('meters', function(t) {
  var km = unit(1000, 'meters').as('km');
  t.is(km, 1);
});

test('kilometers', function(t) {
  var km = unit(1, 'kilometer').as('kilometer');
  t.is(km, 1);
});

test('psi', function(t) {
  var kpa = unit(100, 'psi').as('kpa');
  t.true(approximately(kpa, 689.4744));
});

test('pascals', function(t) {
  var kpa = unit(1000, 'pascals').as('kilopascal');
  t.is(kpa, 1);
});

test('bars', function(t) {
  var kpa = unit(1, 'bar').as('kpa');
  t.is(kpa, 100);
});

test('millibars', function(t) {
  var kpa = unit(100, 'millibars').as('kpa');
  t.is(kpa, 10);
});

test('mmhg', function(t) {
  var kpa = unit(100, 'mmhg').as('kpa');
  t.true(approximately(kpa, 13.3322));
});

test('kilopascals', function(t) {
  var kpa = unit(1, 'kpa').as('kilopascals');
  t.is(kpa, 1);
});

test('kelvins', function(t) {
  var fahrenheit = unit(300, 'kelvins').as('f');
  t.true(approximately(fahrenheit, 80.33));
});

test('fahrenheit', function(t) {
  var celsius = unit(100, 'fahrenheit').as('celsius');
  t.true(approximately(celsius, 37.7778));
});

test('celsius', function(t) {
  var kelvin = unit(100, 'c').as('kelvin');
  t.is(kelvin, 373.15);
});

test('ounces', function(t) {
  var kg = unit(100, 'ounces').as('kg');
  t.true(approximately(kg, 2.83495));
});

test('pounds', function(t) {
  var kg = unit(100, 'lbs').as('kg');
  t.true(approximately(kg, 45.3592));
});

test('grams', function(t) {
  var kg = unit(1000, 'grams').as('kg');
  t.is(kg, 1);
});

test('kilograms', function(t) {
  var kg = unit(1, 'kg').as('kg');
  t.is(kg, 1);
});

test('milliliters', function(t) {
  var l = unit(1000, 'milliliters').as('l');
  t.is(l, 1);
});

test('liters', function(t) {
  var l = unit(1, 'l').as('l');
  t.is(l, 1);
});

test('usgallons', function(t) {
  var l = unit(0.264172, 'usgallons').as('l');
  t.is(l, 1);
});

test('gallons', function(t) {
  var l = unit(0.219969, 'gallons').as('l');
  t.is(l, 1);
});


test/test_index.js
'use strict';

const test = require('ava');
const unit = require('../index');

function approximately(value, expected) {
  return Math.abs(value - expected) <= 0.0001;
}

test('in, as, and to syntaxes are equivalent', function(t) {
  var inches = unit(12, 'inches');
  var toSyntax = inches.to('feet');
  var asSyntax = inches.as('feet');
  var inSyntax = inches.in('feet');
  t.is(toSyntax, inSyntax);
  t.is(toSyntax, asSyntax);
});

test('rounding', function(t) {
  var rounded = unit(13, 'inches').to('feet', 2);
  t.is(rounded, 1.08);
});

test('inches', function(t) {
  var km = unit(100000, 'inches').as('km');
  t.true(approximately(km, 2.54));
});

test('feet', function(t) {
  var km = unit(10000, 'feet').as('km');
  t.true(approximately(km, 3.048));
});

test('miles', function(t) {
  var km = unit(184000, 'miles').as('kilometers');
  t.true(approximately(km, 296119.296));
});

test('meters', function(t) {
  var km = unit(1000, 'meters').as('km');
  t.is(km, 1);
});

test('kilometers', function(t) {
  var km = unit(1, 'kilometer').as('kilometer');
  t.is(km, 1);
});

test('psi', function(t) {
  var kpa = unit(100, 'psi').as('kpa');
  t.true(approximately(kpa, 689.4744));
});

test('pascals', function(t) {
  var kpa = unit(1000, 'pascals').as('kilopascal');
  t.is(kpa, 1);
});

test('bars', function(t) {
  var kpa = unit(1, 'bar').as('kpa');
  t.is(kpa, 100);
});

test('millibars', function(t) {
  var kpa = unit(100, 'millibars').as('kpa');
  t.is(kpa, 10);
});

test('mmhg', function(t) {
  var kpa = unit(100, 'mmhg').as('kpa');
  t.true(approximately(kpa, 13.3322));
});

test('kilopascals', function(t) {
  var kpa = unit(1, 'kpa').as('kilopascals');
  t.is(kpa, 1);
});

test('kelvins', function(t) {
  var fahrenheit = unit(300, 'kelvins').as('f');
  t.true(approximately(fahrenheit, 80.33));
});

test('fahrenheit', function(t) {
  var celcius = unit(100, 'fahrenheit').as('celcius');
  t.true(approximately(celcius, 37.7778));
});

test('celcius', function(t) {
  var kelvin = unit(100, 'c').as('kelvin');
  t.is(kelvin, 373.15);
});

test('ounces', function(t) {
  var kg = unit(100, 'ounces').as('kg');
  t.true(approximately(kg, 2.83495));
});

test('pounds', function(t) {
  var kg = unit(100, 'lbs').as('kg');
  t.true(approximately(kg, 45.3592));
});

test('grams', function(t) {
  var kg = unit(1000, 'grams').as('kg');
  t.is(kg, 1);
});

test('kilograms', function(t) {
  var kg = unit(1, 'kg').as('kg');
  t.is(kg, 1);
});


index.js
'use strict';

var units = require('./units');

var table = {};

Object.keys(units).forEach(function(key) {
  var unit = units[key];
  var aliases = unit.aliases.concat([key]);

  aliases.forEach(function(alias) {
    var base = unit.base;
    table[alias] = {
      to: base ? function(x) { return x / base; } : unit.to,
      from: base ? function(x) { return x * base; } : unit.from,
    };
  });
});

var round = function(n, places) {
  return Number(Math.round(n + 'e+' + places) + 'e-' + places);
};

module.exports = function(value, fromUnit) {
  var methods = {};
  methods.to = function(toUnit, decimalPlaces) {
    var internal = table[fromUnit].to(value);
    var output = table[toUnit].from(internal);
    return decimalPlaces ? round(output, decimalPlaces) : output;
  };
  methods.as = methods.to;
  methods.in = methods.to;
  return methods;
};


units.js
'use strict';

module.exports = {
  in: {
    base: 1000 / 0.3048 * 12,
    aliases: ['inch', 'inches'],
  },
  ft: {
    base: 1000 / 0.3048,
    aliases: ['foot', 'feet'],
  },
  mi: {
    base: 1 / 1.609344,
    aliases: ['mile', 'miles'],
  },
  cm: {
    base: 100000,
    aliases: ['centimeter', 'centimeters'],
  },
  m: {
    base: 1000,
    aliases: ['meter', 'meters', 'metre', 'metres'],
  },
  km: {
    base: 1,
    aliases: ['kilometer', 'kilometers'],
  },
  psi: {
    base: 0.145038,
    aliases: ['psis'],
  },
  pa: {
    base: 1000,
    aliases: ['pascal', 'pascals'],
  },
  bar: {
    base: 0.01,
    aliases: ['bars'],
  },
  mbar: {
    base: 10,
    aliases: ['millibar', 'millibars', 'mbars', 'hectopascal', 'hectopascals'],
  },
  mmhg: {
    base: 7.50061561303,
    aliases: ['torr'],
  },
  atm: {
    base: 0.00986923,
    aliases: ['atmosphere', 'atmospheres', 'atmospheric', 'barometric'],
  },
  kpa: {
    base: 1,
    aliases: ['kpas', 'kilopascal', 'kilopascals'],
  },
  k: {
    from: function(c) { return c + 273.15; },
    to: function(k) { return k - 273.15; },
    aliases: ['kelvin', 'kelvins'],
  },
  f: {
    from: function(c) { return (c * 9 / 5) + 32; },
    to: function(f) { return (f - 32) * 5 / 9; },
    aliases: ['fahrenheit'],
  },
  c: {
    from: function(c) { return c; },
    to: function(c) { return c; },
    aliases: ['celsius', 'centigrade'],
  },
  oz: {
    base: 35.274,
    aliases: ['ounce', 'ounces'],
  },
  lb: {
    base: 2.20462,
    aliases: ['pound', 'lbs', 'pounds'],
  },
  g: {
    base: 1000,
    aliases: ['gm', 'gram', 'grams'],
  },
  kg: {
    base: 1,
    aliases: ['kilogram', 'kilograms'],
  },
  ml: {
    base: 1000,
    aliases: ['milliliter', 'milliliters', 'millilitre', 'millilitres'],
  },
  l: {
    base: 1,
    aliases: ['liter', 'liters', 'litre', 'litres'],
  },
  usgal: {
    base: 0.264172,
    aliases: ['usgallon', 'usgallons'],
  },
  gal: {
    base: 0.219969,
    aliases: ['gallon', 'gallons'],
  },
};
